{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "96d9a4df-84e5-4635-890a-358138fb688d",
   "metadata": {},
   "source": [
    "# Genome Sequencing, Weeks 1 & 2: Genome Assembly\n",
    "\n",
    "In weeks 1 and 2, we address the question of **genome assembly** using **graph algorithms**.\n",
    "\n",
    "We cannot read an organism's genome from start to end, as we do with books. We must first digest the DNA strands into short fragments called **reads**. Then, we must combine the reads into a single genome, in a task called **genome assembly**. The instructors compare genome assembly to blowing up a stack of newspapers with dynamite, then piecing together a complete newspaper from the fragments. \n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/sequencing_overview.png\" width=500px /> \n",
    "</center>\n",
    "\n",
    "But this is not a trivial task! Although bacterial genomes can be pretty small, many organisms have genomes with billions of base pairs: humans have ~3 billion bps, and the fern *Tmesipteris oblanceolata* has a genome of [about 160 billion bps](https://doi.org/10.1016/j.isci.2024.109889). This means that, to sequence a full genome, we have to combine millions of DNA fragments together in the right order. Moreover, DNA is double-stranded, so we cannot know for a given read whether or not we should use it or its reverse complement. The fact that modern sequencing techniques continue to often yield errors means that reads cannot be assembled together with 100% certainty, and some regions may simply fail to be sequenced. \n",
    "\n",
    "The reads generated by modern sequencers often have the same length, so we can assume that all reads have the same length. For now, we will assume that all reads come from the same strand, have no errors, and exhibit perfect coverage, and we'll later relax these assumptions for more realistic datasets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c25349b-dbdf-4ad1-b6f6-823e5d0569ca",
   "metadata": {},
   "source": [
    "## The String Reconstruction Problem\n",
    "\n",
    "The **String Reconstruction** problem is the task of producing, from a collection of reads, a reconstructed string genome. \n",
    "\n",
    "**String Reconstruction Problem**: *Reconstruct a genome from reads.*\n",
    "- **Input**: A collection of strings *Reads*.\n",
    "- **Output**: A string *Genome* reconstructed from *Reads*.\n",
    "\n",
    "But this is not a computational problem—it lacks well-defined conditions that the output *Genome* must satisfy. We will first define the **String Composition** problem, which is the **inverse** of the String Reconstruction problem we eventually would like to solve. Given a string *Text*, we want to give the *k*-mer composition of *Text*, including any repeated *k*-mers.\n",
    "\n",
    "**String Composition Problem**: *Generate the k-mer composition of a string.*\n",
    "- **Input**: An integer *k* and a string *Text*.\n",
    "- **Output**: *Composition*<sub>*k*</sub>(*Text*) (the *k*-mers can be provided in any order).\n",
    "\n",
    "This is trivial. We just iterate through *Text* and return all substrings of length *k*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "010feddf-30b1-45f7-89b2-9b8b71b12be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def composition(k, text):\n",
    "    return [text[i:i+k] for i in range(len(text)-k+1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28c8b301-5e1e-4afc-b241-af890f5e718d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAATC AATCC ATCCA TCCAA CCAAC\n"
     ]
    }
   ],
   "source": [
    "print(\" \".join(composition(5, \"CAATCCAAC\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7992385c-de7a-4b73-8870-7a7c5db2e276",
   "metadata": {},
   "source": [
    "Now, we can revise the String Reconstruction problem to turn it into a computational problem.\n",
    "\n",
    "**String Reconstruction Problem (Revised)**: *Reconstruct a string from its k-mer composition.*\n",
    "- **Input**: An integer *k* and a collection *Patterns* of *k*-mers.\n",
    "- **Output**: A string *Genome* such that *Composition<sub>k</sub>*(*Genome*) is equal to the collection of *k*-mers.\n",
    "\n",
    "This is a more difficult problem to solve. Given a 3-mer composition:\n",
    "\n",
    "    AAT  ATG  ATG  ATG  CAT  CCA  GAT  GCC  GGA  GGG  GTT  TAA  TGC  TGG  TGT\n",
    "\n",
    "The **naive approach** is to start with any 3-mer, say `TAA`, and proceed by chaining it together with an overlapping 3-mer, say `AAT`, and so on: \n",
    "\n",
    "    TAA\n",
    "     AAT\n",
    "      ATG\n",
    "    -----\n",
    "    TAATG\n",
    "\n",
    "At this point, we can consider three 3-mers that begin with `TG` in our composition: `TGC`, `TGG`, and `TGT`. The choice is important: choosing `TGT` means we can only add `GTT` next, and we have no 3-mers that begin with `TT`, so we hit a dead end with `TAATGTT`. \n",
    "\n",
    "\n",
    "    TAA\n",
    "     AAT\n",
    "      ATG\n",
    "       TGT\n",
    "        GTT\n",
    "    -------\n",
    "    TAATGTT\n",
    "\n",
    "\n",
    "\n",
    "With a little bit of thinking ahead, we could reserve `TGT` and `GTT` for the very end of the string. If we extend `ATG` by `TGC` instead, we can continue the process and obtain `TAATGCCATGGATGTT`:\n",
    "\n",
    "\n",
    "    TAA\n",
    "     AAT\n",
    "      ATG\n",
    "       TGC\n",
    "        GCC\n",
    "         CCA\n",
    "          CAT\n",
    "           ATG\n",
    "            TGG\n",
    "             GGA\n",
    "              GAT\n",
    "               ATG\n",
    "                TGT\n",
    "                 GTT\n",
    "    ----------------\n",
    "    TAATGCCATGGATGTT\n",
    "\n",
    "\n",
    "\n",
    "Yet, devilishly, the instructors tell us that this is not the correct string, because we omitted `GGG`! The correct string would be `TAATGCCATGGGATGTT`.\n",
    "\n",
    "The difficulty with choosing between `TGG`, `TGC`, and `TGT` arises because the sequence `ATG` repeats three times in the sequence. The fact that real genomes contain many, many repeats—about 50% of the human genome is made of repeats, of which the [**Alu sequence**](https://en.wikipedia.org/wiki/Alu_element) comprises over a million copies—poses a real challenge to genome assembly. So we need to design a better algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c480cb18-f57e-4224-927a-26427b12e71f",
   "metadata": {},
   "source": [
    "## String Reconstruction as a Walk in the Overlap Graph\n",
    "\n",
    "Repeats in the genome means that we need some way of \"looking ahead\" to choose the right read any time there are multiple options to arrive at a correct reconstruction. One unlikely scenario is one in which we are simply given the sequence information. The collection of *k*-mers *Pattern*, arranged in the correct sequence to produce a string *Genome*, is called the **genome path**. \n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/path_graph.png\" />\n",
    "</center>\n",
    "<br>\n",
    "\n",
    "**String Spelled by a Genome Path Problem**: *Reconstruct a string from its genome path.*\n",
    "- **Input**: A sequence path of *k*-mers *Pattern<sub>1</sub>*, … , *Pattern<sub>n</sub>* such that the last *k*–1 symbols of *Pattern<sub>i</sub>* are equal to the first *k*–1 symbols of *Pattern*<sub>*i*+1</sub> for 1 ≤ *i* ≤ *n*–1.\n",
    "- **Output**: A string *Text* of length *k*+*n*–1 such that the *i*-th *k*-mer in *Text* is equal to *Pattern<sub>i</sub>* (for 1 ≤ *i* ≤ *n*).\n",
    "\n",
    "Given the genome path, it is quite easy to reconstruct *Genome*. There is no need to even check whether the last *k*–1 symbols of *Pattern<sub>i</sub>* are equal to the first *k*–1 symbols of *Pattern*<sub>*i*+1</sub>—we just append, to the first read, the last letter of every subsequent read, since we know that the reads are *k*-mers, and the overlaps are *k*–1 long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cddc47bc-a906-49aa-97d7-5b0b56af27b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def path_to_genome(path):\n",
    "    return path[0] + \"\".join([pattern[-1] for pattern in path[1:]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "102edbf9-648d-4de2-87fb-722a17d8623b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCAGATCATCGG\n"
     ]
    }
   ],
   "source": [
    "path = [\"AGCAGATC\", \"GCAGATCA\", \"CAGATCAT\", \"AGATCATC\", \"GATCATCG\", \"ATCATCGG\"]\n",
    "print(path_to_genome(path))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfe9e090-c38a-4741-b2b0-66eeb62f86f6",
   "metadata": {},
   "source": [
    "But, in real life, we are not given the genome path. Rather, we have to determine the genome path. So, what do we do? Observing that each node on the genome path shares an overlap of length *k*–1 with at least one other node, we can create an **overlap graph** in which, if the last *k*–1 symbols in node A equals the first *k*–1 symbols in node B, we draw an arrow from node A to node B.\n",
    "\n",
    "<br>\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/overlap_graph_lex.png\" width=700px />\n",
    "</center>\n",
    "<br>\n",
    "\n",
    "**Overlap Graph Problem**: *Construct the overlap graph of a collection of k-mers.*\n",
    "- **Input**: A collection *Patterns* of *k*-mers.\n",
    "- **Output**: The overlap graph *Overlap*(*Patterns*), in the form of an adjacency list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "24b78db7-3485-4b2a-9860-166cc09089bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prefix(pattern):\n",
    "    return pattern[:-1]\n",
    "    \n",
    "def suffix(pattern):\n",
    "    return pattern[1:]\n",
    "    \n",
    "def overlap_graph(patterns):\n",
    "    adjacency_list = {}\n",
    "    for p1 in patterns:\n",
    "        adj = [p2 for p2 in patterns if suffix(p1) == prefix(p2)]\n",
    "        if len(adj) != 0:\n",
    "            adjacency_list[p1] = adj\n",
    "    return adjacency_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "169d3a25-5916-4216-81fe-434d47be366f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCATG: CATGC\n",
      "CATGC: ATGCG\n",
      "AGGCA: GGCAT GGCAC\n",
      "GGCAT: GCATG\n"
     ]
    }
   ],
   "source": [
    "patterns = [\"ATGCG\", \"GCATG\", \"CATGC\", \"AGGCA\", \"GGCAT\", \"GGCAC\"]\n",
    "\n",
    "for key, value in overlap_graph(patterns).items():\n",
    "    print(f'{key}: {\" \".join(value)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0163bd-9c5b-4bbb-b89a-94948ca372a2",
   "metadata": {},
   "source": [
    "Then, we can find a candidate for the genome path in the overlap graph by tracing a **Hamiltonian path**, defined as a path that visits each node exactly once (there can be multiple Hamiltonian paths in a single graph).\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/overlap_graph_highlighted.png\" width=700px/>    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "**Hamiltonian Path Problem**: *Construct a Hamiltonian path in a graph.*\n",
    "- **Input**: A directed graph.\n",
    "- **Output**: A path visiting every node in the graph exactly once (if such a path exists).\n",
    "\n",
    "But, the Hamiltonian path problem is NP-complete, meaning there are currently no known algorithms with polynomial (*O*(*n*<sup>*k*</sup>)) or more efficient time complexities. Indeed, if the Hamiltonian path was not highlighted in the overlap graph above, we would have a very difficult time identifying it. We will have to adopt an alternative strategy.\n",
    "\n",
    "## Another Graph for String Reconstruction\n",
    "\n",
    "Ender Nicolaas de Bruijn, a 20th-century Dutch mathematician, who was interested in solving the problem of finding the ***k*-universal string**. Given some length *k*, he wanted to find a string that contains every binary *k*-mer exactly once—that is, a universal string. For example, 0001110100 is a 3-universal string because it contains all eight binary 3-mers, 000, 001, 011, 111, 110, 101, 010, and 100, as substrings, and exactly once. \n",
    "\n",
    "Finding a *k*-universal string, which contains every *k*-mer in a given collection... That sounds a lot like finding a Hamiltonian path, which visits every node given an overlap graph! In fact, the *k*-universal string problem is equivalent to, and can be reduced to, the Hamiltonian path problem. \n",
    "\n",
    "<center>\n",
    "    <img src=\"https://ucarecdn.com/f7ce673d-b1fc-4121-9aeb-9b50666f09e5/\" width=400px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "However, rather than dealing with the intractable Hamiltonian path problem, de Bruijn developed a different way of representing a *k*-mer collection using a graph called **de Bruijn graphs**. Instead of putting each *k*-mer on a node, de Bruijn placed each *k*-mer on an edge, and placed on each node a *k*–1-mer shared by the edges on either side. For example, the below overlap graph:\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/path_graph.png\" width=600px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "is translated into the below de Bruijn graph:\n",
    "\n",
    "<center>\n",
    "    <img src=\"https://ucarecdn.com/63655550-8250-4284-95ab-7c77ddbca891/\" width=600px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "This move, by itself, is not helpful. But, de Bruijn proceeded to \"glue\" or merge identical nodes together:\n",
    "\n",
    "<center>\n",
    "    <img src=\"debruijn.png\" width=400px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "This innovation offers key advantages. Leonhard Euler proved that, in solving the **Bridges of Königsberg** problem, that there exists an efficient, **polynomial**-time algorithm to construct an **Eulerian cycle**—a cycle that travels through every edge in a graph—for every graph, if there exists an Eulerian cycle in said graph. Failing to recognize that a more efficient algorithm existed, for the first two decades following the invention of DNA sequencing, biologists used overlap graphs to assemble genomes, including the human genome. Today, the de Bruijn graph is the dominant approach to genome assembly.\n",
    "\n",
    "Let us first implement the creation of a de Bruijn graph from a string first.\n",
    "\n",
    "**De Bruijn Graph from a String Problem**: *Construct the de Bruijn graph of a string.*\n",
    "- **Input**: An integer *k* and a string *Text*.\n",
    "- **Output**: *DeBruijn*<sub>*k*</sub>(*Text*).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b9d7e9a-a3b2-447d-ab9a-e42be888693b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def de_bruijn_from_string(k, text):\n",
    "    adjacency_list = {}\n",
    "    edges = [text[i:i+k] for i in range(len(text)-k+1)]\n",
    "    for edge in edges:\n",
    "        pref = prefix(edge)\n",
    "        suff = suffix(edge)\n",
    "        adjacency_list.setdefault(pref, []).append(suff)\n",
    "    return adjacency_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "96c2a23e-65f2-4739-8cba-18c9ffd216f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAG: AGA AGA\n",
      "AGA: GAT\n",
      "GAT: ATT\n",
      "ATT: TTC\n",
      "TTC: TCT\n",
      "TCT: CTC CTA\n",
      "CTC: TCT\n",
      "CTA: TAA\n",
      "TAA: AAG\n"
     ]
    }
   ],
   "source": [
    "k = 4\n",
    "text = \"AAGATTCTCTAAGA\"\n",
    "\n",
    "for key, value in de_bruijn_from_string(k, text).items():\n",
    "    print(f'{key}: {\" \".join(value)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7595bee3-011c-4145-8578-791a0935f924",
   "metadata": {},
   "source": [
    "## Walking in the de Bruijn Graph\n",
    "\n",
    "With an overlap graph, we had to find a Hamiltonian path, and that was an intractable problem. But with a de Bruijn graph, we have to find an **Eulerian path**—a path that visits **every edge** rather than every node—to find a candidate for the genome path. \n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/debruijn_graph_eulerian_path.png\" width=400px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "But hold on! In the **de Bruijn Graph from a String** problem, we constructed the de Bruijn graph *given a genome*, but we're trying to reconstruct the genome! Thankfully, we can just as easily construct a de Bruijn graph from isolated reads or *k*-mers. For each *k*-mer, we can create two nodes, suffix and prefix, and assign a directed edge from suffix to prefix. Once we merge all the identical nodes, we'll end up with a linear **composition graph**. Further merging identical nodes in the composition graph results in a de Bruijn graph.\n",
    "\n",
    "<center>\n",
    "    <img src=\"composition.png\" width=600px />    \n",
    "</center>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef25757-c8e0-4aaa-8d6b-b6e0eee2b5e1",
   "metadata": {},
   "source": [
    "The instructors say that this method, creating a linear graph before doing a second merge, will prove useful later.\n",
    "\n",
    "```\n",
    "DeBruijn(Patterns)\n",
    "    dB ← graph in which every k-mer in Patterns is an isolated edge between its prefix and suffix\n",
    "    dB ← graph resulting from gluing all nodes in dB with identical labels\n",
    "    return dB\n",
    "```\n",
    "\n",
    "But creating a de Bruijn graph from a collection of *k*-mers *Pattern* is actually quite easy. Just a small tweak to the previous implementation from the **de Bruijn Graph from a String** problem will do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b6098d3-3f2a-467d-80c4-dceffa5ec4b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def de_bruijn_from_kmers(patterns):\n",
    "    adjacency_list = {}\n",
    "    for p in patterns:\n",
    "        pref = prefix(p)\n",
    "        suff = suffix(p)\n",
    "        adjacency_list.setdefault(pref, []).append(suff)\n",
    "    return adjacency_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "be81d56b-a604-4cb5-95bd-cae0b6c2d6dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAG: AGG\n",
      "CAG: AGG AGG\n",
      "GGG: GGG GGA\n",
      "AGG: GGG\n",
      "GGA: GAG\n"
     ]
    }
   ],
   "source": [
    "patterns = [\"GAGG\", \"CAGG\", \"GGGG\", \"GGGA\", \"CAGG\", \"AGGG\", \"GGAG\"]\n",
    "\n",
    "for key, value in de_bruijn_from_kmers(patterns).items():\n",
    "    print(f'{key}: {\" \".join(value)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e79bc4-5f7a-44f9-8fde-3a2f1c0fbbb7",
   "metadata": {},
   "source": [
    "## Euler's Theorem\n",
    "\n",
    "Week 2 content begins here. \n",
    "\n",
    "As mentioned briefly in Week 1, Euler proved that there exists an efficient, polynomial-time algorithm to construct an Eulerian cycle for every graph, if there exists an Eulerian cycle in said graph. We will examine his algorithm.\n",
    "\n",
    "A graph that possesses an Eulerian cycle is an **Eulerian graph**. And for a graph to be Eulerian, it has to be 1) **balanced** and 2) **strongly connected**. In a **balanced graph**, each node has the same **indegree** (the number of incoming edges) as the **outdegree** (the number of outgoing edges). If an ant were to walk along an Eulerian cycle, if it entered a node through an edge, it should be able to exit through another edge, so the indegree must equal the outdegree. And in a **strongly connected** graph, we have to able to walk from one node to another for all pairs of nodes in the graph. So every Eulerian graph is balanced, but not all balanced graphs are Eulerian, and not all strongly-connected graphs are Eulerian.\n",
    "\n",
    "To prove whether *Graph* has an Eulerian cycle, we can imagine a randomly-walking ant on the graph. Placing the ant on any node and allowing it to walk along only untravelled edges allows it to either find the Eulerian cycle, or get stuck after completing a smaller cycle. For example, the below ant started out at the green node (say *v*<sub>0</sub>), and became stuck after completing a smaller cycle. Once the ant has completed a cycle, we call the cycle *Cycle*<sub>0</sub>.\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/Leo-2.png\" width=300px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "Then, we can start the place the ant on a new node within *Cycle*<sub>0</sub> that still has at least one unused edge, and instruct it to complete a circuit through *Cycle*<sub>0</sub>. Let us call the new node *v*<sub>1</sub>. Once the ant returns to *v*<sub>1</sub>, it will find an unused edge to escape *Cycle*<sub>0</sub> and continue its journey. \n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/Leo-3.png\" width=300px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "If the ant becomes stuck again, like below, we can name *Cycle*<sub>1</sub> the newly completed cycle (which includes *Cycle*<sub>0</sub>).\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/Leo-4.png\" width=300px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "And we repeat the procedure, placing the ant on a node in *Cycle*<sub>1</sub> with an unused edge, until the ant succeeds in traversing all edges.\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/Leo_final_steps.png\" width=600px />    \n",
    "</center>\n",
    "<br>\n",
    "\n",
    "## From Euler's Theorem to an Algorithm for Finding Eulerian Cycles\n",
    "\n",
    "This is a **constructive proof**, which not only proves the result but also provides us with an algorithm. Let us implement Euler's proof.\n",
    "\n",
    "```\n",
    "EulerianCycle(Graph)\n",
    "    form a cycle Cycle by randomly walking in Graph (don't visit the same edge twice!)\n",
    "    while there are unexplored edges in Graph\n",
    "        select a node newStart in Cycle with still unexplored edges\n",
    "        form Cycle’ by traversing Cycle (starting at newStart) and then randomly walking \n",
    "        Cycle ← Cycle’\n",
    "    return Cycle\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3c7701cb-b79e-46c9-ac75-f13e5ff0a396",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def eulerian_cycle(graph):\n",
    "    \n",
    "    def random_walk(graph, current_node):\n",
    "        '''\n",
    "        Complete a cycle by walking through an Eulerian graph randomly.\n",
    "        '''\n",
    "        cycle = [current_node]\n",
    "        while True:\n",
    "            if current_node in graph.keys(): # if we're not stuck\n",
    "                next_node = random.choice(graph[current_node])\n",
    "                cycle.append(next_node)\n",
    "    \n",
    "                # remove new edge to be traversed from future consideration\n",
    "                if len(graph[current_node]) > 1:\n",
    "                    graph[current_node].remove(next_node)\n",
    "                else:\n",
    "                    graph.pop(current_node)\n",
    "                    \n",
    "                current_node = next_node\n",
    "            else:\n",
    "                break\n",
    "        return cycle, graph\n",
    "\n",
    "    def prev_cycle(cycle, new_start):\n",
    "        '''\n",
    "        Once new start node is selected, shift previous cycle to begin\n",
    "        with new start node. \n",
    "        '''\n",
    "        new_start_index = cycle.index(new_start) # all nodes are unique so this is fine\n",
    "        return cycle[new_start_index:] + cycle[1:new_start_index+1]\n",
    "    \n",
    "    current_node = random.choice(list(graph.keys()))    \n",
    "    cycle, graph = random_walk(graph, current_node)\n",
    "\n",
    "    while len(graph) > 0:\n",
    "        # choose new node to start from, given the previous iteration's cycle. \n",
    "        # if node is **left** in graph.keys(), i.e. still has unused outgoing edge(s)\n",
    "        new_start = random.choice([node for node in cycle if node in graph.keys()])\n",
    "        \n",
    "        prev = prev_cycle(cycle, new_start)\n",
    "        new, graph = random_walk(graph, new_start)\n",
    "        cycle = prev + new[1:]\n",
    "    \n",
    "    return cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1bd27c39-dfc1-42d8-86f0-b9e38cf836bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 5 4 2 1 0 3 2 6 8 7 9 6\n"
     ]
    }
   ],
   "source": [
    "graph = {0: [3], 1: [0], 2: [1, 6], 3: [2], 4: [2], 5: [4], 6: [5, 8], 7: [9], 8: [7], 9: [6]}\n",
    "print(\" \".join(str(x) for x in eulerian_cycle(graph)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "956b9325-0e0c-44ba-9873-a0214c5eca5f",
   "metadata": {},
   "source": [
    "We can now check whether a directed graph has an Eulerian *cycle*, but not an Eulerian *path*. This is |straightforwardly remedied. A nearly balanced graph has an Eulerian path if and only if adding an edge between its unbalanced nodes makes the graph balanced and strongly connected. So, we can **simply connect the two unbalanced nodes** to form an Eulerian cycle.\n",
    "\n",
    "<center>\n",
    "    <img src=\"http://bioinformaticsalgorithms.com/images/Assembly/Eulerian_path_to_cycle.png\" width=600px />    \n",
    "</center>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "58eaecc0-bae0-4932-8c65-e605d67652f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_unbalanced(graph):\n",
    "    '''\n",
    "    Find unbalanced nodes in a nearly balanced graph with a Eulerian path.\n",
    "    '''\n",
    "    sources = graph.keys()\n",
    "    dests = sum(list(graph.values()), []) # list of arrow destinations\n",
    "    all_nodes = set(list(sources) + list(dests))\n",
    "    odds_dict = {}\n",
    "    for node in all_nodes:\n",
    "        if node in sources:\n",
    "            n = len(graph[node]) # number of outdegrees\n",
    "        else:\n",
    "            n = 0\n",
    "        for node2 in dests:\n",
    "            if node == node2:\n",
    "                n -= 1\n",
    "        if n % 2 != 0:\n",
    "            odds_dict[n] = node\n",
    "\n",
    "    return odds_dict[1], odds_dict[-1]\n",
    "    \n",
    "def eulerian_path(graph):\n",
    "    orig, term = find_unbalanced(graph)\n",
    "    graph.setdefault(term, []).append(orig) # add temporary edge to graph\n",
    "\n",
    "    path = eulerian_cycle(graph)[:-1]\n",
    "\n",
    "    # slice Eulerian cycle at temporary edge\n",
    "    for i in range(len(path)-1):\n",
    "        if path[0] == orig and path[len(path)-1] == term:\n",
    "            return path\n",
    "        if [path[i], path[i+1]] == [term, orig]:\n",
    "            return path[i+1:] + path[0:i+1]\n",
    "    \n",
    "    return path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "940ac055-bc37-4ae2-8274-7272fc1f99a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 7 8 9 6 3 0 2 1 3 4\n"
     ]
    }
   ],
   "source": [
    "data = {0: [2], 1: [3], 2: [1], 3: [0, 4], 6: [3, 7], 7: [8], 8: [9], 9: [6]}\n",
    "print(\" \".join([str(x) for x in eulerian_path(data)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35de6708-555e-4456-88da-f322949f9f20",
   "metadata": {},
   "source": [
    "Somewhat unceremoniously, we have now arrived at a solution to the **String Reconstruction** problem. Given a collection of reads *Patterns*, we first create the de Bruijn graph, then we create the Eulerian path, and then we string together the Eulerian path into a single string *Text*.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4e5048bc-62da-413f-8fbc-a87f8799cd9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def string_reconstruction(patterns):\n",
    "    de_bruijn_graph = de_bruijn_from_kmers(patterns)\n",
    "    path = eulerian_path(de_bruijn_graph)\n",
    "    return path_to_genome(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cba81e8f-3243-421e-8397-1766300bf4de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GGCTTACCA'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patterns = [\"CTTA\", \"ACCA\", \"TACC\", \"GGCT\", \"GCTT\", \"TTAC\"]\n",
    "\n",
    "string_reconstruction(patterns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acef550f-5cb8-4898-a377-fa01860bb91c",
   "metadata": {},
   "source": [
    "Recall that de Bruijn's goal was to find the *k*-universal binary string for any given *k*. The instructors tell us that de Bruijn was actually interested in finding *k*-universal *circular* strings. This is quite similar to the problem of assembling a circular genome, like those of most bacteria. We can easily use the **Eulerian Cycle** algorithm to answer this question. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ff0b8da0-64f4-4218-866b-92383b9388a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# binary strings function taken from StackOverflow: \n",
    "# https://stackoverflow.com/questions/64890117/what-is-the-best-way-to-generate-all-binary-strings-of-the-given-length-in-pytho\n",
    "def generate_binary_strings(bit_count):\n",
    "    binary_strings = []\n",
    "    def genbin(n, bs=''):\n",
    "        if len(bs) == n:\n",
    "            binary_strings.append(bs)\n",
    "        else:\n",
    "            genbin(n, bs + '0')\n",
    "            genbin(n, bs + '1')\n",
    "\n",
    "    genbin(bit_count)\n",
    "    return binary_strings\n",
    "\n",
    "def k_universal_circular_string(k):\n",
    "    patterns = generate_binary_strings(k)\n",
    "    de_bruijn_graph = de_bruijn_from_kmers(patterns)\n",
    "    path = eulerian_cycle(de_bruijn_graph)\n",
    "    return path_to_genome(path[:-(k-1)]) # I don't know why we need the k-1 term, but sure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5b8ade3b-82a8-4c67-bb5c-9313aeb9e5aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1110011010000101\n"
     ]
    }
   ],
   "source": [
    "print(k_universal_circular_string(4))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b0c8ac-9efc-435e-8d07-c3116256c070",
   "metadata": {},
   "source": [
    "## Assembling Genomes from Read-Pairs\n",
    "\n",
    "Generate the (3,2)-mer composition of TAATGCCATGGGATGTT in lexicographic order. Include repeats, and return your answer as a list on a single line.  As a hint to help you with formatting, your answer should begin \"(AAT|CAT) (ATG|ATG)...\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7a797319-e021-449e-98a3-df7e3ea49055",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_pairs(k, d, text):\n",
    "    pairs = []\n",
    "    for i in range(len(text)-2*k-d+1):\n",
    "        pairs.append([text[i:i+k], text[i+k+d:i+2*k+d]])\n",
    "    pairs.sort()\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8dcccda7-8bff-4624-b778-1173d7590abe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(AAT|CAT) (ATG|ATG) (ATG|ATG) (CAT|GAT) (CCA|GGA) (GCC|GGG) (GGG|GTT) (TAA|CCA) (TGC|TGG) (TGG|TGT)\n"
     ]
    }
   ],
   "source": [
    "text = \"TAATGCCATGGGATGTT\"\n",
    "print(\" \".join([f'({pair[0]}|{pair[1]})' for pair in read_pairs(3, 2, text)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da266c7f-ffeb-437e-906d-c8c2f3361578",
   "metadata": {},
   "source": [
    "**String Reconstruction from Read-Pairs Problem**: *Reconstruct a string from its paired composition.*\n",
    "- **Input**: A collection of paired *k*-mers *PairedReads* and an integer *d*.\n",
    "- **Output**: A string *Text* with (*k*,*d*)-mer composition equal to *PairedReads* (if such a string exists).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b51ea43c-a68d-4b0c-8270-dda2ea68f18e",
   "metadata": {},
   "source": [
    "```\n",
    "StringSpelledByGappedPatterns(GappedPatterns, k, d)\n",
    "    FirstPatterns ← the sequence of initial k-mers from GappedPatterns\n",
    "    SecondPatterns ← the sequence of terminal k-mers from GappedPatterns\n",
    "    PrefixString ← StringSpelledByPatterns(FirstPatterns, k)\n",
    "    SuffixString ← StringSpelledByPatterns(SecondPatterns, k)\n",
    "    for i = k + d + 1 to |PrefixString|\n",
    "        if the i-th symbol in PrefixString does not equal the (i - k - d)-th symbol in SuffixString\n",
    "            return \"there is no string spelled by the gapped patterns\"\n",
    "    return PrefixString concatenated with the last k + d symbols of SuffixString\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fd004ead-c711-44f4-bf70-a109fa6b47f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GACCGAGCGCCGGA'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def string_from_gapped_patterns(gapped_patterns, k, d):\n",
    "    first_patterns = [p.split(\"|\")[0] for p in gapped_patterns]\n",
    "    second_patterns = [p.split(\"|\")[1] for p in gapped_patterns]\n",
    "    prefix_string = path_to_genome(first_patterns)\n",
    "    suffix_string = path_to_genome(second_patterns)\n",
    "    for i in range(k+d+1, len(prefix_string)):\n",
    "        if prefix_string[i] != suffix_string[i-k-d]:\n",
    "            return \"No string spelled by gapped patterns\"\n",
    "    return prefix_string + suffix_string[-(k+d):]\n",
    "    \n",
    "k = 4\n",
    "d = 2\n",
    "patterns = ['GACC|GCGC', 'ACCG|CGCC', 'CCGA|GCCG', 'CGAG|CCGG', 'GAGC|CGGA']\n",
    "\n",
    "string_from_gapped_patterns(patterns, k, d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "81602809-0e0b-4203-b7d6-20e73df14867",
   "metadata": {},
   "outputs": [],
   "source": [
    "patterns = [\"GAGA|TTGA\", \"TCGT|GATG\", \"CGTG|ATGT\", \"TGGT|TGAG\", \"GTGA|TGTT\", \"GTGG|GTGA\", \"TGAG|GTTG\", \"GGTC|GAGA\", \"GTCG|AGAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3e91ee5a-a9f6-44ee-ac29-38a40fbe2bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def de_bruijn_from_paired_kmers(patterns):\n",
    "    adjacency_list = {}\n",
    "    for p in patterns:\n",
    "        pair = p.split(\"|\")\n",
    "        pref = f'{prefix(pair[0])}|{prefix(pair[1])}'\n",
    "        suff = f'{suffix(pair[0])}|{suffix(pair[1])}'\n",
    "        adjacency_list.setdefault(pref, []).append(suff)\n",
    "    return adjacency_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d03c22fa-4039-474f-9d50-0cd74403bb88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GAG|TTG': ['AGA|TGA'],\n",
       " 'TCG|GAT': ['CGT|ATG'],\n",
       " 'CGT|ATG': ['GTG|TGT'],\n",
       " 'TGG|TGA': ['GGT|GAG'],\n",
       " 'GTG|TGT': ['TGA|GTT'],\n",
       " 'GTG|GTG': ['TGG|TGA'],\n",
       " 'TGA|GTT': ['GAG|TTG'],\n",
       " 'GGT|GAG': ['GTC|AGA'],\n",
       " 'GTC|AGA': ['TCG|GAT']}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_bruijn_from_paired_kmers(patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4491eea3-f6b1-40c2-a544-8aaeddc207fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
